File: 1247-decrease-elements-to-make-array-zigzag/decrease-elements-to-make-array-zigzag.py
class Solution:
    def movesToMakeZigzag(self, nums: List[int]) -> int:
        n = len(nums)
        evenans = 0
        for i in range(0,n,2):
            if i==0 and n>1:
                if nums[0] >= nums[1]:
                    evenans += nums[0] - nums[1] + 1
                continue
            if i==n-1:
                if nums[i] >= nums[i-1]:
                    evenans += nums[i] - nums[i-1] + 1
                continue
            if nums[i] >= nums[i-1]:
                if nums[i] >= nums[i+1]:
                    evenans += nums[i] - min(nums[i-1], nums[i+1]) + 1
                    continue
                else:
                    evenans += nums[i] - nums[i-1] + 1
            else:
                if nums[i] >= nums[i+1]:
                    evenans += nums[i] - nums[i+1] + 1
        oddans = 0    
        for i in range(1,n,2):
            if i==n-1:
                if nums[i] >= nums[i-1]:
                    oddans += nums[i] - nums[i-1] + 1
                continue
            if nums[i] >= nums[i-1]:
                if nums[i] >= nums[i+1]:
                    oddans += nums[i] - min(nums[i-1], nums[i+1]) + 1
                    continue
                else:
                    oddans += nums[i] - nums[i-1] + 1
            else:
                if nums[i] >= nums[i+1]:
                    oddans += nums[i] - nums[i+1] + 1

        return min(evenans, oddans)

File: 1247-decrease-elements-to-make-array-zigzag/decrease-elements-to-make-array-zigzag.cpp
class Solution {
public:
    int movesToMakeZigzag(vector<int>& nums) {
        int n=nums.size();
        int costOdd=0;
        vector<int>temp=nums;
        for(int i=1;i<n;i+=2){
            if(temp[i-1]>=temp[i]){
                costOdd=costOdd+(temp[i-1]-temp[i])+1;
                temp[i-1]=temp[i]-1;
            }
            if(i<n-1 && temp[i+1]>=temp[i]){
                costOdd=costOdd+(temp[i+1]-temp[i])+1;
                temp[i+1]=temp[i]-1;
            }
        }
        int costEven=0;
        temp=nums;
        for(int i=0;i<n;i+=2){
            if(i>0 && temp[i-1]>=temp[i]){
                costEven=costEven+(temp[i-1]-temp[i])+1;
                temp[i-1]=temp[i]-1;
            }
            if(i<n-1 && temp[i+1]>=temp[i]){
                costEven=costEven+(temp[i+1]-temp[i])+1;
                temp[i+1]=temp[i]-1;
            }
        }
        int ans=min(costEven,costOdd);
        return ans;
    }
};